### PR TITLE
Swap test for testing state overlaps

### PR DESCRIPTION
**Pull request by ncrubin**
Naive, uncompiled, swap test using standard gates in pyquil.

Updates should involve 1) removal of number of shots in exchange for a
confidence interval the user is willing to tolerate 2) A compiled
version of CSWAP for Rigetti's hardware.

### CODE REVIEW

**Comment by mpharrigan** [comment]
Do you want to add some docs + references about what the swap test is?

File: grove/circuit_primitives/swap.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+Given two states existing on registers A and B, their overlap can be measured
+by performing a swap test.
+"""
+import numpy as np
+from pyquil.quil import Program
+from pyquil.gates import H, CSWAP
+
+
+class RegisterSizeMismatch(Exception):
+    """Error associated with mismatch register size"""
```
Comment (mpharrigan) [inline grove/circuit_primitives/swap.py]: mismatched

File: grove/circuit_primitives/swap.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+def swap_circuit_generator(register_a, register_b, ancilla):
+    """
+    Generate the swap test circuit primitive.
+
+    Registers A and B must be of equivalent size for swap to work.  This module
+    uses the CSWAP gate in pyquil.
+
+    :param List register_a: qubit labels in the 'A' register
+    :param List register_b: qubit labels in the 'B' register
+    :param ancilla: ancilla to measure and control of swap operation.
```
Comment (mpharrigan) [inline grove/circuit_primitives/swap.py]: control of -> control the?

File: grove/circuit_primitives/swap.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    :return: Program
+    """
+    if len(register_a) != len(register_b):
+        raise RegisterSizeMismatch("registers involve different numbers of qubits")
+
+    if not isinstance(register_a, list):
+        raise TypeError("Register A need to be list")
```
Comment (mpharrigan) [inline grove/circuit_primitives/swap.py]: needs

**Comment by ncrubin** [comment]
I will add more docs when the circuit primitives module is filled out.

**Pull request merged**